frontend/userForm.tsx

```diff
@@ -0,0 +1,306 @@
+                          ];
+/**
+ * Provides a form for changing user metadata and user roles. The form will
+ * load data for the user id `id` if it's a number, or for a new user if `id` is
+ * `undefined` or `'new'`.
+ */
+export function UserForm({id}: {id: number | 'new' | undefined}) {
+  const {genebanks, loadData} = useDataContext()
+  const [user, setUser] = React.useState({...defaultValues} as ManagedUser)
+  const [isNew, setNew] = React.useState(false)
+  const [level, setLevel] = React.useState(PermissionLevels[0])
+  const [genebank, setGenebank] = React.useState(genebanks.length > 0 ? {value: genebanks[0].id, label: genebanks[0].name} : null)
+  const [herd, setHerd] = React.useState(null as any)
+  const history = useHistory()
+  const classes = useStyles();
+
+  const loadUser = (id: number) => {
+    get(`/api/manage/user/${id}`).then(
+      data => {
+        if (data) {
+          setUser(data)
+          setNew(false)
+        } else {
+          setNew(true)
+        }
+      },
+      error => console.error(error)
+    );
+  }
+
+  /**
+   * Loads the user information for `id`, or sets the form to accept a new user.
+   */
+  React.useEffect(() => {
+    setUser({...defaultValues})
+    if (typeof id == 'number') {
+      loadUser(id)
+      setNew(false)
+    } else if (id == 'new' || !id) {
+      setNew(true)
+    }
+  }, [id])
+
+  /**
+   * If `isNew` is set, this function creates a POST request to create a new
+   * user in the backend, and then navigates to edit the new user. Otherwise
+   * the current form is sent as an UPDATE request to update the user `id`.
+   */
+  const submitForm = () => {
+    let postData = {...user};
+    delete postData["privileges"];
+    let protocol = isNew ? post : update;
+    protocol(`/api/manage/user/${id == 'new' ? 0 : id}`, postData).then(
+      data => {
+        switch (data.status) {
+          case "updated": console.info("updated."); break; // updated user
+          case "success":
+            history.push(`/manage/user/${data.id}`);
+            setNew(false);
+            break; // added user
+          default: console.warn("status:", data)// "failed" or other erro
+        }
+      },
+      error => console.error(error)
+    )
+  }
+
+  /**
+   * Returns a list of all herds of the genebank identified by `genebankId`,
+   * formatted as react-select options, i.e. `{value: <id>, label: <name>}`.
+   *
+   * @param genebankId Id of the genebank to get herds from.
+   */
+  const genebankHerds = (genebankId: number | undefined) => {
+    if (genebankId === undefined || genebank === null) {
+      return []
+    }
+
+    const currentGenebank = genebanks.find((g: Genebank) => g.id == genebank.value);
+    if (currentGenebank === undefined) {
+      return []
+    }
+    return currentGenebank.herds.map((h: Herd) => {
+      return {value: h.id, label: `${h.herd}${h.herd_name ? ` - ${h.herd_name}` : ''}`}
```

Comment:
Perhaps make a helper function of this label pattern (which is surisingly hard to read with a ternary nested in a template string)

Reply:
yeah - I can't really say why I just kept re-writing it instead of finally making a function :D